File: packages/camera/camera_avfoundation/example/ios/RunnerTests/CameraCaptureSessionQueueRaceConditionTests.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

@import camera_avfoundation;
@import XCTest;
@import AVFoundation;
#import <OCMock/OCMock.h>

@interface CameraCaptureSessionQueueRaceConditionTests : XCTestCase
@end

@implementation CameraCaptureSessionQueueRaceConditionTests

- (void)testFixForCaptureSessionQueueNullPointerCrashDueToRaceCondition {
  SwiftCameraPlugin *camera = [[SwiftCameraPlugin alloc] initWithRegistry:OCMProtocolMock(@protocol(FlutterTextureRegistry)) messenger:OCMProtocolMock(@protocol(FlutterBinaryMessenger))];


  id mockDevice = OCMClassMock([AVCaptureDevice class]);
  OCMStub([mockDevice authorizationStatusForMediaType:OCMOCK_ANY])
      .andReturn(AVAuthorizationStatusAuthorized);

  XCTestExpectation *disposeExpectation =
      [self expectationWithDescription:@"dispose's result block must be called"];
  XCTestExpectation *createExpectation =
      [self expectationWithDescription:@"create's result block must be called"];
  FlutterMethodCall *disposeCall = [FlutterMethodCall methodCallWithMethodName:@"dispose"
                                                                     arguments:nil];
  FlutterMethodCall *createCall = [FlutterMethodCall
      methodCallWithMethodName:@"create"
                     arguments:@{@"resolutionPreset" : @"medium", @"enableAudio" : @(1)}];
  // Mimic a dispose call followed by a create call, which can be triggered by slightly dragging the
  // home bar, causing the app to be inactive, and immediately regain active.
  [camera handleMethodCall:disposeCall
                    result:^(id _Nullable result) {
                      [disposeExpectation fulfill];
                    }];

  [camera handleMethodCall:createCall result:^(id  _Nullable result) {
    [createExpectation fulfill];
  }];

  [self waitForExpectationsWithTimeout:1 handler:nil];
  // `captureSessionQueue` must not be nil after `create` call. Otherwise a nil
  // `captureSessionQueue` passed into `AVCaptureVideoDataOutput::setSampleBufferDelegate:queue:`
  // API will cause a crash.
  XCTAssertNotNil(camera.test_captureSessionQueue,
                  @"captureSessionQueue must not be nil after create method. ");
}

@end

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/CameraCaptureSessionQueueRaceConditionTests.swift
//
//  CameraCaptureSessionQueueRaceConditionTests.swift
//  RunnerTests
//
//  Created by Huan Lin on 6/29/22.
//  Copyright © 2022 The Flutter Authors. All rights reserved.
//

import Foundation
import Flutter
import XCTest
@testable import camera_avfoundation;

final class CameraCaptureSessionQueueRaceConditionTests: XCTestCase {
  func testFixForCaptureSessionQueueNullPointerCrashDueToRaceCondition() {
    let cam = SwiftCameraPlugin(registry: MockTextureRegistry(), messenger: MockBinaryMessenger())

    MockPermissionService.authorizationStatusForMediaTypeStub = { _ in .authorized }

    let disposeExpectation = expectation(description: "dispose's result block must be called")
    let createExpectation = expectation(description: "create's result block must be called")
    let disposeCall = FlutterMethodCall(methodName: "dispose", arguments: nil)
    let createCall = FlutterMethodCall(methodName: "create", arguments: ["resolutionPreset": "medium", "enableAudio": NSNumber(booleanLiteral: true)])


    // Mimic a dispose call followed by a create call, which can be triggered by slightly dragging the
    // home bar, causing the app to be inactive, and immediately regain active.

    cam.handle(disposeCall) { _ in
      disposeExpectation.fulfill()
    }

    cam.handle(createCall) { _ in
      createExpectation.fulfill()
    }

    waitForExpectations(timeout: 1)

    // `captureSessionQueue` must not be nil after `create` call. Otherwise a nil
    // `captureSessionQueue` passed into `AVCaptureVideoDataOutput::setSampleBufferDelegate:queue:`
    // API will cause a crash.
    
    XCTAssertNotNil(cam.test_captureSessionQueue, "captureSessionQueue must not be nil after create method. ")



  }
}
